### PR TITLE
Re-implement the `put`, `replace`, and `extend` operators as documented

### PR DESCRIPTION
**Pull request by dominiklohmann**
This new version of `put` the operator projects, re-orders, and also supports extractors and selectors besides fixed values.

Here's an example:

```
❯ gunzip -c vast/integration/data/json/sip.log.json.gz \
  | vast exec 'read json | head 1 | put user_agent, schema=#type, time=#import_time, first_ip=:ip | write json' \
  | jq
{
  "user_agent": "friendly-scanner",
  "schema": "b751f83ded9bfc9d",
  "time": "1970-01-01T00:00:00.000000",
  "first_ip": "119.57.72.26"
}
```

The logic has also been implemented for `extend` and `replace` accordingly.

Fixes tenzir/issues#286

### CODE REVIEW

**Comment by dominiklohmann** [comment]
> The `#type` and `#import_time` extractors don't seem to work yet.

They do work correctly, but the example has an inferred type so `#type` is the type fingerprint `b751f83ded9bfc9d` and the data was never imported, so `#import_time` is the UNIX epoch.

Arguably for the import time, the function should add nulls if the import time is not available.

**Comment by mavam** [comment]
While you're touching this code, would it make sense to add the `#node_id` selector as well?

**Comment by dominiklohmann** [comment]
> While you're touching this code, would it make sense to add the `#node_id` selector as well?

I'd say no, but we can do so pretty soon in a follow-up PR. We have this tracked on the roadmap, and this PR is neither the time nor the place for it.

**Comment by jachris** [comment]
When I tested this, `replace` with type-extractors (as in `replace :ip=123)` did not work, presumably because your PR still uses `resolve_key_suffix` to resolve the extractor.

**Comment by dominiklohmann** [comment]
> @jachris When I tested this, `replace` with type-extractors (as in `replace :ip=123)` did not work, presumably because your PR still uses `resolve_key_suffix` to resolve the extractor.

Right, that is just missing. I have a fix locally but that is rather complicated. Would you prefer that to be in another PR or still pushed on top of this one?

File: web/docs/understand/operators/transformations/replace.md [hunk 1]
```diff
@@ -15,10 +15,10 @@ The `replace` operator mutates existing fields by providing a new value.
 The difference between `replace` and [`extend`](extend.md) is that `replace`
 overwrites existing fields, whereas `extend` doesn't touch the input.
 
-### `<field=operand>`
+### `<extractor=operand>`
 
-The assignment consists of `field` that describes an existing field name and
-`operand` that defines the new field value.
+The assignment consists of an `extractor` that describe match existing fields
+and `operand` that defines the new field value.
```
Comment (jachris) [inline web/docs/understand/operators/transformations/replace.md:21]:
```suggestion
The assignment consists of an `extractor` that matches against existing fields
and an `operand` that defines the new field value.
```

**Review by jachris — APPROVED** [review]
Very nice :+1:

File: libvast/src/table_slice.cpp [hunk 2]
```diff
@@ -718,4 +718,109 @@ table_slice resolve_enumerations(table_slice slice) {
...
+        VAST_ASSERT(append_result.ok(), append_result.ToString().c_str());
+      }
+      array = builder->Finish().ValueOrDie();
+    };
+    caf::visit(g, inferred_type);
+  };
+  // Helper function that binds an existing array.
+  auto bind_array = [&](const offset& index) {
+    inferred_type = layout.field(index).type;
+    array = static_cast<arrow::FieldPath>(index).Get(*batch).ValueOrDie();
```
[the remark this answers is not in the document]
Comment (dominiklohmann) [reply libvast/src/table_slice.cpp:780]: No particular reason.

**Comment by jachris** [comment]
This should be fixed with 59f1689.

**Pull request merged**